### PR TITLE
Improvements to the variants API

### PR DESCRIPTION
**Pull request by david4096**
Most noticeable when the call list is large.

I got at worst a 2x speedup. When an empty callset list is sent you'll see much faster but YMMV.

Replaced the hacked phase set implementation with pysam native. Found that pysam and pyvcf disagree on whether a single genotype in a variant set should be called phased. See demonstration [here](https://gist.github.com/david4096/997fc3ecb00783ba6775acf15750eec4).

Helps with #950

### CODE REVIEW

File: ga4gh/datamodel/variants.py [hunk 1]
```diff
@@ -123,6 +124,7 @@ def addCallSet(self, sampleName):
         self._callSetIdMap[callSetId] = callSet
         self._callSetNameMap[sampleName] = callSet
         self._callSetIds.append(callSetId)
+        self._callSets[callSet.getId()] = len(self._callSetIds) - 1
```
Comment (diekhans) [inline ga4gh/datamodel/variants.py]: The name _callSets implies that this contains CallSet objects and is confusing. maybe _callSetIdToIndex?

File: ga4gh/datamodel/variants.py [hunk 2]
```diff
@@ -685,15 +687,19 @@ def convertVariant(self, record, callSetIds):
...
+                    call = protocol.Call()
+                    call.callSetId = callSetId
+                    call.callSetName = callSet.getSampleName()
+                    index = self._callSets[callSetId]
+                    genotypeData = sampleData[index].split(
+                             ":")[0]  # REMOVAL
+                    call.genotype, call.phaseset = convertVCFGenotype(
+                             genotypeData, call.phaseset)
+                    variant.calls.append(call)
         variant.id = self.getVariantId(variant)
```
Comment (diekhans) [inline ga4gh/datamodel/variants.py]: +1 very important, sans comments above.

We still need to seriously look at performance.

File: ga4gh/datamodel/variants.py [hunk 3]
```diff
@@ -685,15 +687,19 @@ def convertVariant(self, record, callSetIds):
...
+                if callSet is not None:
+                    call = protocol.Call()
+                    call.callSetId = callSetId
+                    call.callSetName = callSet.getSampleName()
+                    index = self._callSets[callSetId]
+                    genotypeData = sampleData[index].split(
+                             ":")[0]  # REMOVAL
+                    call.genotype, call.phaseset = convertVCFGenotype(
+                             genotypeData, call.phaseset)
+                    variant.calls.append(call)
```
Comment (diekhans) [inline ga4gh/datamodel/variants.py]: I am not a coder on this, so feel free to ignore; however I find it much easier to read code if functions are created rather than nesting loops inside of ifs

File: ga4gh/datamodel/variants.py [hunk 4]
```diff
@@ -625,36 +597,25 @@ def _addDataFile(self, filename):
...
-
-        ###########################################
-        # call.genotype = list(pysamCall.allele_indices)
-        # call.phaseset = pysamCall.phaseset
-        ###########################################
-
-        call.genotypeLikelihood = []
+    def _convertGaCall(self, callSet, pysamCall):
+        phaseset = None
+        if pysamCall.phased is True:
```
Comment (jeromekelleher) [inline ga4gh/datamodel/variants.py]:
```python
if pysamCall.phased:
     # stuff
```

File: ga4gh/datamodel/variants.py [hunk 5]
```diff
@@ -680,20 +641,14 @@ def convertVariant(self, record, callSetIds):
...
-        sampleIterator = 0  # REMOVAL
-        if callSetIds is not None:
-            for name, call in record.samples.iteritems():
-                if self.getCallSetId(name) in callSetIds:
-                    genotypeData = sampleData[sampleIterator].split(
-                        ":")[0]  # REMOVAL
-                    variant.calls.append(self._convertGaCall(
-                        record.id, name, call, genotypeData))  # REPLACE
-                sampleIterator += 1  # REMOVAL
+        if callSetIds is not None and len(callSetIds) > 0:
```
Comment (jeromekelleher) [inline ga4gh/datamodel/variants.py]: When could ``callSetIds`` be None here? I don't think that's a valid input. We can just ``assert`` that callSetIds is not None I think.

File: ga4gh/datamodel/variants.py [hunk 6]
```diff
@@ -680,20 +641,14 @@ def convertVariant(self, record, callSetIds):
...
+            for callSetId in callSetIds:
+                callSet = self.getCallSet(callSetId)
```
Comment (jeromekelleher) [inline ga4gh/datamodel/variants.py]: Why would callSet be None here? This seems like an error condition to me. We shouldn't silently hide error conditions, this will cause very hard to find bugs. Much better to fail loudly with an exception.

**Comment by jeromekelleher** [comment]
This is a much needed cleanup, thanks @david4096. I think we can simplify things a bit by identifying our error cases more clearly and raising exceptions when nasty things happen. 

Re the pysam/pyvcf issue, can you give us some examples of where they disagree?

**Comment by david4096** [comment]
@jeromekelleher the one example I found in our data you look at here https://gist.github.com/david4096/997fc3ecb00783ba6775acf15750eec4

**Comment by jeromekelleher** [comment]
I'm not sure what to make of this phasing issue @david4096 --- this looks like an error in the input vcf to me, since we have '0' rather than '0/0', '0/1' etc in the genotype column, unlike the rest of the file. I could be wrong though. In any case, it seems like it's an awkward corner case, and I'm happy to assume that pysam has got it right. So, I would work around this in the tests by:
```python
if len(genotype.split("/")) == 1:
     # assert phased is true
else:
     # Compare our value for phased with pyvcf
```

How does this sound?

**Comment by david4096** [comment]
@jeromekelleher Works for me! Updated!

**Comment by jeromekelleher** [comment]
Excellent, thanks @david4096. This is a super important update, I'm excited to see how our benchmarks look after we merge.

+1.

**Comment by jeromekelleher** [comment]
Can you rebase this please @david4096? We want to get this in ASAP.

**Comment by david4096** [comment]
Sure updated!

**Comment by jeromekelleher** [comment]
Looks like we're out of date again. Can you rebase and merge please @david4096?
